src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,414 @@
+
+        """
+        last_return = None
+        for _ in runner.step_epochs():
+            for _ in range(self.steps_per_epoch):
+                if not self._buffer_prefilled:
+                    batch_size = int(self.min_buffer_size)
+                else:
+                    batch_size = None
+                runner.step_path = runner.obtain_samples(
+                    runner.step_itr, batch_size)
+                path_returns = []
+                for path in runner.step_path:
+                    self.replay_buffer.add_transitions(
+                        observation=path['observations'],
+                        action=path['actions'],
+                        reward=path['rewards'],
+                        next_observation=path['next_observations'],
+                        terminal=path['dones'])
+                    path_returns.append(sum(path['rewards']))
+                assert len(path_returns) is len(runner.step_path)
+                self.episode_rewards.append(np.mean(path_returns))
+                for _ in range(self._gradient_steps):
+                    policy_loss, qf1_loss, qf2_loss = self.train_once()
+            last_return = log_performance(runner.step_itr,
+                                          self._obtain_evaluation_samples(
+                                              runner.get_env_copy(),
+                                              num_trajs=10),
+                                          discount=self.discount)
+            self._log_statistics(policy_loss, qf1_loss, qf2_loss)
+            tabular.record('TotalEnvSteps', runner.total_env_steps)
+            runner.step_itr += 1
+
+        return np.mean(last_return)
+
+    def train_once(self, itr=None, paths=None):
+        """Complete 1 training iteration of SAC.
+
+        Args:
+            itr (int): Iteration number. This argument is deprecated.
+            paths (list[dict]): A list of collected paths.
+                This argument is deprecated.
+
+        Returns:
+            torch.Tensor: loss from actor/policy network after optimization.
+            torch.Tensor: loss from 1st q-function after optimization.
+            torch.Tensor: loss from 2nd q-function after optimization.
+
+        """
+        del itr
+        del paths
+        if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:
+            samples = self.replay_buffer.sample(self.buffer_batch_size)
+            samples = tu.dict_np_to_torch(samples)
+            policy_loss, qf1_loss, qf2_loss = self.optimize_policy(0, samples)
+            self._update_targets()
+
+        return policy_loss, qf1_loss, qf2_loss
+
+    def _get_log_alpha(self, **kwargs):
+        """Return the value of log_alpha.
+
+        This function exists in case there are versions of sac that need
+        access to a modified log_alpha, such as multi_task sac.
+
+        Args:
+            kwargs(dict): keyword args that can be used in retrieving the
```

Comment:
kwargs never get passed, so can you just delete it? Is it going to be used somewhere?

Reply:
its going to be used in MTSAC.

Reply:
Ah, I see. That makes sense.